### PR TITLE
feat: add table lock manager to fix table lock cannot be released

### LINKED ISSUE #13239: bug: If first exec update return err, the next update will return TableAlreadyLocked.
**Summary**

```
create table testc(c35 int);

insert into testc values(1);


> UPDATE testc SET c35 = (12498456046895769003 !) WHERE to_timestamp(8472945773481);
ERROR 1105 (HY000): Internal. Code: 1001, Text = eval try eval const failed:
unable to cast type `Timestamp` to type `Boolean NULL`.
> UPDATE testc SET c35 = (12498456046895769003 !) WHERE to_timestamp(8472945773481);
ERROR 1105 (HY000): TableAlreadyLocked. Code: 2015, Text = table is locked by other session, please retry later.
```

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Add table lock manager to fix table lock cannot be released bug.

Add setting `acquire_lock_timeout` to set the maximum timeout in seconds for acquire a lock, the default value is 10 sec.
Move `table-lock` out of the ee features.

```
mysql> create table testc(c35 int);
Query OK, 0 rows affected (0.06 sec)

mysql> insert into testc values(1);
Query OK, 1 row affected (0.08 sec)

mysql> UPDATE testc SET c35 = (12498456046895769003 !) WHERE to_timestamp(8472945773481);
ERROR 1105 (HY000): BadArguments. Code: 1006, Text = eval try eval const failed:
unable to cast type `Timestamp` to type `Boolean NULL`.
mysql> UPDATE testc SET c35 = (12498456046895769003 !) WHERE to_timestamp(8472945773481);
ERROR 1105 (HY000): BadArguments. Code: 1006, Text = eval try eval const failed:
unable to cast type `Timestamp` to type `Boolean NULL`.
```

- Closes #13239 

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13434)

### CODE REVIEW

File: src/query/storages/fuse/src/operations/mutation/compact/segment_compact_mutator.rs [hunk 1]
```diff
@@ -137,11 +138,19 @@ impl SegmentCompactMutator {
         // summary of snapshot is unchanged for compact segments.
         let statistics = self.compact_params.base_snapshot.summary.clone();
         let fuse_table = FuseTable::try_from_table(table.as_ref())?;
-        let handler = TableLockHandlerWrapper::instance(self.ctx.clone());
-        let mut heartbeat = handler
-            .try_lock(self.ctx.clone(), fuse_table.table_info.clone())
+
+        let lock_mgr = TableLockManagerWrapper::instance(self.ctx.clone());
```
Comment (zhang2014) [inline src/query/storages/fuse/src/operations/mutation/compact/segment_compact_mutator.rs:142]: We can define api by standard library mutex style?

File: src/query/ee/src/table_lock/table_lock_manager.rs [hunk 2]
```diff
@@ -66,12 +104,8 @@ impl TableLockHandler for RealTableLockHandler {
             }
 
             // Get the previous revision, watch the delete event.
-            let lock_key = TableLockKey {
-                table_id,
-                revision: reply[position - 1],
-            };
             let req = WatchRequest {
-                key: lock_key.to_string_key(),
+                key: lock.watch_delete_key(reply[position - 1]),
```
Comment (zhang2014) [inline src/query/ee/src/table_lock/table_lock_manager.rs:108]: Maybe dead lock when lock key is expired?

File: src/query/ee-features/table-lock/src/table_level_lock/lock.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+use common_meta_app::schema::TableLockKey;
+use common_meta_kvapi::kvapi::Key;
+use common_pipeline_core::TableLock;
+use common_pipeline_core::TableLockLevel;
+use common_pipeline_core::TableLockReq;
+
+use crate::table_level_lock::req::*;
+use crate::TableLockManagerWrapper;
+
+pub struct TableLevelLock {
```
Comment (zhang2014) [inline src/query/ee-features/table-lock/src/table_level_lock/lock.rs:26]: maybe move it to oss mod is better

File: src/query/ee/src/table_lock/table_lock_manager.rs (hunk 2 above)
Comment (zhyass) [reply src/query/ee/src/table_lock/table_lock_manager.rs:108]: https://github.com/datafuselabs/databend/blob/6fc5bb2783f40d4cd6cc9e8a30317c046ffba073/src/meta/service/tests/it/grpc/metasrv_grpc_watch.rs#L324 
Also can get the delete event when key is expired.

**Review by lichuang — APPROVED** [review]

File: src/query/pipeline/core/src/pipeline.rs [hunk 4]
```diff
@@ -54,6 +55,7 @@ pub struct Pipeline {
     pub pipes: Vec<Pipe>,
     on_init: Option<InitCallback>,
     on_finished: Option<FinishedCallback>,
+    lock_guards: Vec<Arc<LockGuard>>,
```
Comment (zhang2014) [inline src/query/pipeline/core/src/pipeline.rs:58]:
```suggestion
    lock_guards: Vec<LockGuard>,
```

**Review by zhang2014 — APPROVED** [review]

**Comment by zhyass** [comment]
askbend:summary

**Comment by drmingdrmer** [comment]
What is the identifier of a table lock?
Is that table ID plus revision or just the table ID?
In other words, Is it possible for two lock the same table id but with different revisions to be acquired?

**Comment by zhyass** [comment]
> What is the identifier of a table lock? Is that table ID plus revision or just the table ID? In other words, Is it possible for two lock the same table id but with different revisions to be acquired?

It's table ID plus revision. Here is the previous pr https://github.com/datafuselabs/databend/pull/11238
```
__fd_table_lock/table_id/revision -> TableLockMeta
```

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 5 of 34 files at r1, 1 of 3 files at r2, 2 of 36 files at r5, 7 of 49 files at r6, 19 of 56 files at r7, 7 of 19 files at r9.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13434)* status: 35 of 93 files reviewed, 7 unresolved discussions (waiting on @dantengsky, @soyeric128, @Xuanwo, @zhang2014, and @zhyass)

___
*[`src/meta/api/src/schema_api_impl.rs` line 3260 at r9](https://reviewable.io/reviews/datafuselabs/databend/13434#-Ni2ZAZ571PNPJHtUe_f:-Ni2ZAZ571PNPJHtUe_g:undefined) ([raw file](https://github.com/datafuselabs/databend/blob/5103a6cb31f642a92e2807dabd26fc83c550dde3/src/meta/api/src/schema_api_impl.rs#L3260)):*
> ```XML
>             if req.acquire_lock {
>                 table_lock_meta.acquired_on = Some(Utc::now());
>             }
> ```





_Suggestion:_
```XML
            // Set `acquire_lock = true` to initialize `acquired_on` when the
            // first time this lock is acquired. Before the lock is
            // acquired(becoming the first in lock queue), or after being
            // acquired, this argument is always `false`.
            if req.acquire_lock {
                table_lock_meta.acquired_on = Some(Utc::now());
            }
```
___
*[`src/query/catalog/src/lock_api.rs` line 51 at r9](https://reviewable.io/reviews/datafuselabs/databend/13434#-Ni2avPg3OXqzI5PqHlm:-Ni2avPg3OXqzI5PqHln:b-q26du3) ([raw file](https://github.com/datafuselabs/databend/blob/5103a6cb31f642a92e2807dabd26fc83c550dde3/src/query/catalog/src/lock_api.rs#L51)):*
> ```XML
>     async fn check_lock(&self, catalog: Arc<dyn Catalog>) -> Result<bool> {
>         let list_table_lock_req = self.list_table_lock_req();
>         let reply = catalog.list_table_lock_revs(list_table_lock_req).await?;
> ```

Why isn't listing table lock revisions a function of the ListApi?
But instead it's a method of the catalog?
___
*[`src/query/service/src/catalogs/default/database_catalog.rs` line 624 at r9](https://reviewable.io/reviews/datafuselabs/databend/13434#-Ni2cQZ51LGbIfoylJKZ:-Ni2cQZ51LGbIfoylJK_:b-81s6ro) ([raw file](https://github.com/datafuselabs/databend/blob/5103a6cb31f642a92e2807dabd26fc83c550dde3/src/query/service/src/catalogs/default/database_catalog.rs#L624)):*
> ```XML
>     async fn create_table_lock_rev(
>         &self,
>         req: Box<dyn LockRequest>,
> ```

What's the necessity to use a trait object instead of a concrete type for lock request?
IMHO, the lock request has only one implementation in our codebase, and its fields are all fixed.

**Comment by zhyass** [comment]
askbend:summary

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 1 of 28 files at r12, 1 of 36 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13434)* status: 24 of 98 files reviewed, 7 unresolved discussions (waiting on @dantengsky, @soyeric128, @Xuanwo, @zhang2014, and @zhyass)

___
*[`src/meta/app/src/schema/lock.rs` line 39 at r13](https://reviewable.io/reviews/datafuselabs/databend/13434#-Ni7u3Wq9ccJNTZN3Xwn:-Ni7u3Wq9ccJNTZN3Xwo:b-n5pgxe) ([raw file](https://github.com/datafuselabs/databend/blob/5dd304725a0b1cf99d23f2ff04b5439b37c10b62/src/meta/app/src/schema/lock.rs#L39)):*
> ```XML
>     /// table-level lock.
>     Table,
> }
> ```

How about we rename `LockLevel` to `LockKey` and have it include `table_id` and
`revision`. This way, it would be more intuitive to generate a key or a prefix.



_Suggestion:_
```XML
#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum LockLevel {
    /// table-level lock.
    Table{ table_id:u64, revision: u64},
}
```

File: src/meta/proto-conv/tests/it/proto_conv.rs [hunk 5]
```diff
@@ -464,6 +477,16 @@ fn test_build_pb_buf() -> anyhow::Result<()> {
...
     }
 
+    // LockMeta
+    {
+        let table_lock_meta = new_lock_meta();
+        let p = table_lock_meta.to_pb()?;
+
+        let mut buf = vec![];
+        common_protos::prost::Message::encode(&p, &mut buf)?;
+        println!("table_lock_meta:{:?}", buf);
```
Comment (zhang2014) [inline src/meta/proto-conv/tests/it/proto_conv.rs:487]:
```suggestion
```

File: src/query/storages/common/locks/src/lock_manager.rs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+use crate::LockHolder;
+
+pub struct LockManager {
+    active_locks: Arc<RwLock<HashMap<u64, Arc<Mutex<LockHolder>>>>>,
+    tx: Sender<u64>,
+}
+
+impl LockManager {
+    pub fn init() -> Result<()> {
+        let (tx, rx) = async_channel::unbounded();
```
Comment (zhang2014) [inline src/query/storages/common/locks/src/lock_manager.rs:51]: need record metric for channel queue size

File: src/query/storages/common/locks/src/lock_manager.rs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        let active_locks = Arc::new(RwLock::new(HashMap::new()));
+        let lock_manager = Self { active_locks, tx };
+        GlobalIORuntime::instance().spawn({
+            let active_locks = lock_manager.active_locks.clone();
+            async move {
+                while let Ok(revision) = rx.recv().await {
+                    let lock = active_locks.write().remove(&revision);
```
Comment (zhang2014) [inline src/query/storages/common/locks/src/lock_manager.rs:58]:
```suggestion
                    if let Some(lock) = active_locks.write().remove(&revision) {}
```

File: src/query/storages/common/locks/src/lock_manager.rs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+                    if let Some(lock) = lock {
+                        let mut guard = lock.lock().await;
```
Comment (zhang2014) [inline src/query/storages/common/locks/src/lock_manager.rs:60]: Maybe use `std:: sync:: metux` is better?

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 5 of 20 files at r4, 15 of 49 files at r6, 6 of 56 files at r7, 5 of 28 files at r12, 15 of 36 files at r13, 15 of 22 files at r14, 19 of 20 files at r16, 1 of 1 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13434)* status: all files reviewed, 10 unresolved discussions (waiting on @dantengsky, @soyeric128, @Xuanwo, @zhang2014, and @zhyass)

**Review by zhang2014 — APPROVED** [review]

**Pull request merged**